### ISSUE TITLE
Add a simple example about how to build your own C++ Pyodide package

### ISSUE THREAD

**Issue by nyckmaia** [pyodide/pyodide#2711]
## 📚 Documentation

I saw that **there is already** a page on the Pyodide [website](https://pyodide.org/en/stable/development/new-packages.html) titled _"Creating a Pyodide package"_, but for a `C/C++` developer it **is not clear how to proceed.**

In my conversation [here](https://github.com/pyodide/pyodide/issues/2588#issuecomment-1153937476), @JeanChristopherMorinPerso informed me that it is possible to create Pyodide packages from `C++` code that has the `Pybind11` wrapper and that he made it himself for [his project](https://github.com/AcademySoftwareFoundation/OpenTimelineIO).

My `C++` library **already ha**s the `Pybind11` wrapper, but even so, **I'm not finding documentation or a simple example that I can follow and get a valid Pyodide package.**

Finally, I tried to understand what was the reasoning behind compiling Pyodide packages which were originally written in `C/C++` contained on Github in the '[packges](https://github.com/pyodide/pyodide/tree/main/packages)' folder, (example: '[numpy](https://github.com/pyodide/pyodide/tree/main/packages/numpy)'), but **I couldn't understand the logic of the contents of these folders** and how everything is actually working.

So for libraries written in `C/C++` I have **2 suggestions:**

Suggestion 01: C++ code without Pybind11 wrapper
----------------------------------------------------------
Adding to the documentation (maybe even in the `README`) is a simple little example of **how to create a Pyodide package from a `C++` library**.
This simple example could be named: `Hello World`.

**Example:**

```
#include <iostream>

class HelloWorld
{
public:
    HelloWorld() {
        std::cout << "Hello" << std::endl;
    }
}
```

Suggestion 02: Code with Pybind11 wrapper
-------------------------------------------------
Add the same type of simple example mentioned above to the documentation but that contains a `Pybind11` wrapper

**Comment by hoodmane** [comment]
Yeah it would be great to have more documentation for this.

> of how to create a Pyodide package from a C++ library

Are you looking for something like `packages/sharedlib-test` and `packages/sharedlib-test-py`?

**Comment by nyckmaia** [comment]
> Are you looking for something like `packages/sharedlib-test` and `packages/sharedlib-test-py`?

Thanks @hoodmane , that would be almost it.

Part 01
-------
These 2 examples you gave me are great, but they are for the **C language**, that is, without classes and other C++ tools.

It would be great to have a similar example, but **involving a C++ class and a Standard Library cal**l like `std::cout` for example, as I mentioned above in the suggestions.

**Is there already something along these lines?**

Part 02
-------
Even though you can make the wrapper manually, as shown in the 2 examples you mentioned above, **having the possibility to automate this process is extremely advantageous and attractive.**

I don't know how this _"automated wrapper"_ works, but I believe that the use of `Pybind11` should help this process, since this is its role and that @JeanChristopherMorinPerso has [already said](https://github.com/pyodide/pyodide/issues/2588#issuecomment-1152813742) that it managed to create a Pyodide package through `Pybind11`.

I truly believe that a very simple example involving a `C++` class with a `Pybind11` wrapper will entice many `C++` developers to use Pyodide.

**Would it be possible for you to create this example?**

Thanks!

**Comment by josephrocca** [comment]
I'm a newbie here too but am happy to try help create these docs. @nyckmaia Could you provide a simple "hello world" Python package that builds and works fine in Python, and then we can try to piece together a simple Pyodide build process based on that and put it into a new doc page?

This is the process I've been using for other packages per the existing `new-package.md` docs:

```bash
git clone https://github.com/pyodide/pyodide && cd pyodide
./run_docker --pre-built # this mounts current directory as /src
make  # this takes several minutes, unfortunately
pip install ./pyodide-build
python -m pyodide_build mkpkg <package-name>
cd packages/<package-name>
python -m pyodide_build buildpkg meta.yaml
```

But due to the packages that I'm trying to build, I always run into several issues that require updates to the generated meta.yaml file. Can you try publishing your hello world to PyPi and then running the above to see where you run into problems?

@hoodmane I notice that `python -m pyodide_build buildpkg meta.yaml` has been replaced with `python -m pyodide_build buildall --only 'package-name' packages dist` in `new-package.md`, but the latter doesn't seem to work? It seems to just hang and not output anything. Is that new command a work-in-progress? It would be good to get the latest docs published to pyodide.org soon I think, since it seems like a lot of people are trying to use the current docs, and they're a bit confusing.

**Comment by nyckmaia** [comment]
Hi @josephrocca 

I didn't understood exactly what I should do.

**Question 01:**
Did you mean that I should create a new package inside `Pyodide` folder containing pure `C++` code? As my "Hello World" above?

**Question 02:**
If yes, should I write the `Pybind11` wrapper too? Or do you prefer the only pure `C++` class?

**Question 03:**
How can I inform the build command for this "Hello World" project?
Example without `CMake`:
``` 
g++ -Wall -c ./src/mylib.cpp -I ./include
ar rvs libmylib.a mylib.o
```

**Question 04:**
Can I use `CMake`?

**Comment by ryanking13** [comment]
> **Question 04:**
> Can I use `CMake`?

Yes, there are some packages that use CMake. For example, `opencv-python` uses `scikit-build` to use cmake inside setuptools, and `xgboost` calls cmake directly via subprocess during the build.

In order to build packages that use CMake, you'll have to write [a custom toolchain file to set variables correctly](https://github.com/pyodide/pyodide/blob/main/packages/xgboost/cmake/Toolchain.cmake), which is not very comfortable for new contributors.

I have some plans to mitigate this situation by automatically using the CMake toolchain file that Emscripten provides, but I wasn't able to make it work until now. So for now, I would say that it is not easy for new contributors to add a package that uses CMake.

**Comment by hoodmane** [comment]
> it is not easy for new contributors to add a package that uses CMake

Nor is it easy for me. My approach is to ask henryiii or ryanking13.

**Comment by juangburgos** [comment]
+1 for a simple example, after some patches, I managed to build [pyarma](https://gitlab.com/jason-rumengan/pyarma) using `emcmake cmake ...` and finally obtained a `pyarma.cpython-312-x86_64-linux-gnu.so` file which containes the package, but now how do I load it into pyodide? 

I tried zipping it into a `.whl` file, serve it on `http` and load it with `micropip` but that does not seem to work. Any advise? Thanks!

**Comment by ryanking13** [comment]
@juangburgos

Could you try `pyodide build` (see: https://pyodide.org/en/stable/development/building-and-testing-packages.html)?

**Comment by juangburgos** [comment]
@ryanking13 thanks for replying
> Could you try `pyodide build`

That is the first thing I tried, but always get:
```
ModuleNotFoundError: No module named '_sysconfigdata__emscripten_wasm32-emscripten'
```

I then tried `emcmake cmake ..` which at least builds something.

BTE, if it would work, is it not `pyodide build --exports pyinit` ? I read that somewhere.

**Comment by henryiii** [comment]
FYI, if you want to see examples in CI, NumPy, boost-histogram, and awkward all build pyodide wheels in ci. https://github.com/scikit-hep/awkward/blob/main/.github/workflows/docs.yml Is the most complex, but it makes the wheel available in the live docs. I think scikit-image does this too.

**Comment by hoodmane** [comment]
> it makes the wheel available in the live docs

Cool!

**Comment by pthom** [comment]
I think that the option 

```cmake
pyodide build --exports pyinit
```
should be mentioned clearly in this page: https://pyodide.org/en/stable/development/building-and-testing-packages.html#building-and-testing-packages-out-of-tree

I guess this is the recommended option when building with pybind.

I discovered this option after a long deep dive into the internals of pyodide-build where my link failed because it was exporting all the functions of my library and causing a weird error inside Python's subprocess (command line too long, because the command line did include exports for 10000+ functions)

**Comment by hoodmane** [comment]
I wonder if we should also change the default behavior to `--export pyinit`. It's the default for in tree builds.

**Comment by ryanking13** [comment]
I actually thought that `--export pyinit` it was a default for out-of-tree too. Anyway, we need to fix the command line too long error for calculating exports.
